csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,474 @@
+    ValGraph* graph) {
+  AbstractTensorSchedule::apply(abstract_tensor, tvs, graph);
+}
+
+AbstractTensor forwardAroundMissingAxes(
+    const AbstractTensor& abstract_tensor,
+    TensorView* tv) {
+  ValGraph* graph = nullptr;
+
+  NVF_ERROR(abstract_tensor.size() > 0);
+  const AbstractId& first_abs_id = abstract_tensor.domain.front();
+  NVF_ERROR(first_abs_id.is<ValGroupAndItsGraph>());
+  graph = first_abs_id.as<ValGroupAndItsGraph>().graph;
+
+  // We will need to map the ValGroups in abstract_tensor to new ones. We do
+  // this recursively using this map. We initialize it by mapping all the loop
+  // ValGroups in tv to themselves. When we traverse the graph in topological
+  // order we update this mapping for each ExprGroup whose input ValGroups are
+  // all found in the map. If a ValGroup is not in the map that means it is not
+  // computable. If we pass a merge operation and only one ValGroup is
+  // computable, we forward that computable ValGroup by updating this mapping.
+  std::unordered_map<ValGroup, ValGroup> computable_valgroup_map;
+
+  for (IterDomain* id : tv->getLoopDomain()) {
+    ValGroup vg = graph->toGroup(id);
+    computable_valgroup_map.emplace(vg, vg);
+  }
+
+  // TODO: ideally we could use something like
+  // getExprsTo(abstract_tensor.domain) in case there are ValGroups downstream
+  // of abstract_tensor
+  ExprGroups expr_groups = ValGraphStmtSort(*graph).exprs();
+  for (const ExprGroup& eg : expr_groups) {
+    const std::vector<ValGroup> input_groups = graph->inputGroups(eg);
+
+    std::vector<ValGroup> computable_valgroups;
+    for (const ValGroup& vg : input_groups) {
+      auto it = computable_valgroup_map.find(vg);
+      if (it != computable_valgroup_map.end()) {
+        computable_valgroups.push_back(it->second);
+      }
+    }
+
+    if (computable_valgroups.empty()) {
+      // This ExprGroup is not computable or forwardable
+      continue;
+    }
+
+    const std::vector<ValGroup> output_groups = graph->outputGroups(eg);
+
+    if (computable_valgroups == input_groups) {
+      // All input groups are computable as-is since their ValGroups match, so
+      // this ExprGroup is computable as-is
+      for (const ValGroup& vg : output_groups) {
+        computable_valgroup_map.emplace(vg, vg);
+      }
+    } else if (computable_valgroups.size() == input_groups.size()) {
+      // All input groups are computable but some of them are different than
+      // the originals due to previous forwarding.
+      std::vector<ValGroup> computable_output_groups =
+          replayExprGroup(graph, eg, computable_valgroups);
+      NVF_ERROR(computable_output_groups.size() == output_groups.size());
+      for (size_t i : c10::irange(output_groups.size())) {
+        computable_valgroup_map.emplace(
+            output_groups[i], computable_output_groups[i]);
+      }
+    } else if (computable_valgroups.size() == 1 && eg->front()->isA<Merge>()) {
+      // Forward the input group
```

Comment:
I'm still not comfortable here. Why is it ok to forward whatever missing merge inputs?

Reply:
This is the type of behavior I'm trying to enable here:
```c++
  TensorView* tv0 = makeSymbolicTensor(1);
  TensorView* tv1 = makeSymbolicTensor(2);
  fusion.addInput(tv0);
  fusion.addInput(tv1);

  TensorView* tv2 = mul(tv0, IrBuilder::create<Val>(3.0));
  TensorView* tv3 = broadcast(tv2, {false, true});
  TensorView* tv4 = mul(tv3, tv1);

  fusion.addOutput(tv4);

  tv4->merge(0);
  tv4->split(0, 32);
  tv4->axis(-1)->parallelize(ParallelType::TIDx);
  tv4->axis(-2)->parallelize(ParallelType::BIDx);

  scheduler_utils::transformPropagateToAllFrom(tv4, -1);
  scheduler_utils::parallelizeAllLike(tv4);

  inlineMost();

  fusion.printMath();
  fusion.printTransforms();
```
```
Inputs:
  T0_g[ iS19{( ceilDiv(i0, 32) )}, iS20{32} ], float
  T1_g[ iS12{( ceilDiv(( i2 * i3 ), 32) )}, iS13{32} ], float
Outputs:
  T4_g[ iblockIdx.x9{( ceilDiv(( i0 * i3 ), 32) )}, ithreadIdx.x10{32} ] ca_pos( 2 ) produce_pos( 2 ), float

%kernel_math {
T2_l[ iblockIdx.x17{( ceilDiv(i0, 32) )}, ithreadIdx.x18{32} ] ca_pos( 2 )
   = T0_g[ iS19{( ceilDiv(i0, 32) )}, iS20{32} ]
   * double(3);
T3_l[ iblockIdx.x15{( ceilDiv(( i0 * 1 ), 32) )}, ithreadIdx.x16{32} ] ca_pos( 2 ) produce_pos( 2 )
   = broadcast( T2_l[ iblockIdx.x17{( ceilDiv(i0, 32) )}, ithreadIdx.x18{32} ] ca_pos( 2 ) )
T4_g[ iblockIdx.x9{( ceilDiv(( i0 * i3 ), 32) )}, ithreadIdx.x10{32} ] ca_pos( 2 ) produce_pos( 2 )
   = T3_l[ iblockIdx.x15{( ceilDiv(( i0 * 1 ), 32) )}, ithreadIdx.x16{32} ] ca_pos( 2 ) produce_pos( 2 )
   * T1_g[ iS12{( ceilDiv(( i2 * i3 ), 32) )}, iS13{32} ];
}

T0_g[ iS19{( ceilDiv(i0, 32) )}, iS20{32} ]
 logical domain : (iS0{i0})
 contiguity: f
  Split: iS0{i0} by factor 32 -> iS19{( ceilDiv(i0, 32) )}, iS20{32}
 loop domain : (iS19{( ceilDiv(i0, 32) )}, iS20{32})
T2_l[ iblockIdx.x17{( ceilDiv(i0, 32) )}, ithreadIdx.x18{32} ] ca_pos( 2 )
 logical domain : (iS3{i0})
 contiguity: t
  Split: iS3{i0} by factor 32 -> iblockIdx.x17{( ceilDiv(i0, 32) )}, ithreadIdx.x18{32}
 loop domain : (iblockIdx.x17{( ceilDiv(i0, 32) )}, ithreadIdx.x18{32})
T3_l[ iblockIdx.x15{( ceilDiv(( i0 * 1 ), 32) )}, ithreadIdx.x16{32} ] ca_pos( 2 ) produce_pos( 2 )
 logical domain : (iS4{i0}, bS5{1})
 contiguity: t n
  Merge: iS4{i0} and bS5{1} -> iS14{( i0 * 1 )}
  Split: iS14{( i0 * 1 )} by factor 32 -> iblockIdx.x15{( ceilDiv(( i0 * 1 ), 32) )}, ithreadIdx.x16{32}
 loop domain : (iblockIdx.x15{( ceilDiv(( i0 * 1 ), 32) )}, ithreadIdx.x16{32})
T1_g[ iS12{( ceilDiv(( i2 * i3 ), 32) )}, iS13{32} ]
 logical domain : (iS1{i2}, iS2{i3})
 contiguity: f f
  Merge: iS1{i2} and iS2{i3} -> iS11{( i2 * i3 )}
  Split: iS11{( i2 * i3 )} by factor 32 -> iS12{( ceilDiv(( i2 * i3 ), 32) )}, iS13{32}
 loop domain : (iS12{( ceilDiv(( i2 * i3 ), 32) )}, iS13{32})
T4_g[ iblockIdx.x9{( ceilDiv(( i0 * i3 ), 32) )}, ithreadIdx.x10{32} ] ca_pos( 2 ) produce_pos( 2 )
 logical domain : (iS6{i0}, iS7{i3})
 contiguity: t t
  Merge: iS6{i0} and iS7{i3} -> iS8{( i0 * i3 )}
  Split: iS8{( i0 * i3 )} by factor 32 -> iblockIdx.x9{( ceilDiv(( i0 * i3 ), 32) )}, ithreadIdx.x10{32}
 loop domain : (iblockIdx.x9{( ceilDiv(( i0 * i3 ), 32) )}, ithreadIdx.x10{32})
```
Although we start with a merge on `tv4`, we "ignore" it when propagating back to `tv2`, and we still replay the `Split`. The "forwarding around missing axes" is just that bit of using the existing input to stand in for the merged axis.

Reply:
Yes, but that forwarding is not done arbitrary but is only done for some specific cases as defined by `ForwardingInfo`. My concern is in this case it seems it just ignores any missing merge inputs.

Reply:
When I read through the `ForwardingInfo` code [here](https://github.com/NVIDIA/Fuser/blob/main/csrc/transform_iter.cpp#L723-L832) it seemed like it was just doing something equivalent to what I implemented here: it forwards axes that are squeezed or broadcast through Merge expressions.

Reply:
In `ForwardingInfo` the propagation is done only across `BroadcastOp` and `SqueezeOp` expressions. I think this is actually how the Permissive graph is built in `IdModel`, by checking `ForwardingInfo` for each expression. So in that sense the logic is still in effect since "missing" in the `AbstractTensor` case corresponds to "not permissive mapped into the given ValGroup".

Reply:
I think you're right, but it seems this could expand the definition of the Permissive graph. Let's say:

```
// tv0: [I0]
auto tv0 = makeSymbolicTensor(1); // input
// tv1: [I0, I1]
auto tv1 = makeSymbolicTensor(2); // input
// tv0_cache: [I0]
auto tv0_cache = set(tv0);
// tv1_cache: [I0, I1]
auto tv1_cache = set(tv1);
// tv2: [I0, R1]
auto tv2 = sum(tv1_cache, {1});
// tv3: [I0]
auto tv3 = add(tv0_cache, tv2); // output

tv1_cache->merge(0, 1)->split(0, 4);
// tv1_cache: [I0*I1/4, 4]
```

Here, `ForwardingInfo` would not do anything as there's no broadcast, so the `I0*I1` domain of `tv1_cache` would be mapped neither with `I0` or `I1`, which would mean `tv0_cache` wouldn't be inlined into `tv1_cache`. In other words, the current transform propagator wouldn't propagate the merge and split of `tv1_cache` to `tv0_cache`. That would be fine since `tv0_cache` wasn't used for `tv1_cache`.

Now, with this AbstractTensor scheduling, if an AbstractTensor consists of the loop domains of `tv1_cache` and if it's applied to `tv0_cache`, IIUC, it would transform `tv0_cache` by skipping the merge and applying the split. The resulting Loop graph would be also updated accordingly.

To me, this seems like a subtle but important difference compared to the current behavior. I'm not agains the new behavior. I'm just trying to understand the difference as clearly as possible.

Let me know if I'm missing anything.

Reply:
I think you're right and reductions are where we might have a difference since that is the other place besides squeeze where we might remove an axis.

It's true that currently I am updating the graph by inserting the new IterDomain when it's created. I can make that optional. I will think more on the effect of examples like this.

In the matmul case I think I don't hit this because we are not merging reduction groups with non reduction groups and propagating past the reduction. In that case I am only using the graph for this purpose so corrupting it is not a big concern. This is one reason I avoid rebuilding the IdModel after starting scheduling (another is that the ExprGroups introduced by scheduling an AbstractTensor are lost when the graph is rebuilt).

Reply:
Note that the above example does not merge reduction and non-reduction domains.

I think it makes most sense to start with the least change from the current behavior and explore extensions as next steps. In particular, can we drop `forwardAroundMissingAxes` for now and would that still be enough for the mamtul fusion?

Reply:
You're right that the example doesn't merge reduction and non-reduction domains directly; I meant that if we used `applyAbstractTransforms` in this case that we would do that merge/split on `tv2` and we would forward around the merge and apply the split on `tv0_cache`/`tv3`. This is indeed a difference with this approach compared to propagation.

BTW I tried your example with propagation and unless you explicitly stop propagating at the reduction it will also error out by trying to merge the reduction/iteration domains in `tv2`. We actually hit the exact same error message in either case:
```c++
TEST_F(AbstractTensorTest, TestApplyScheduling2) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  // tv0: [I0]
  auto tv0 = makeSymbolicTensor(1); // input
  // tv1: [I0, I1]
  auto tv1 = makeSymbolicTensor(2); // input
  fusion.addInput(tv0);
  fusion.addInput(tv1);
  // tv0_cache: [I0]
  auto tv0_cache = set(tv0);
  // tv1_cache: [I0, I1]
  auto tv1_cache = set(tv1);
  // tv2: [I0, R1]
  auto tv2 = sum(tv1_cache, {1});
  // tv3: [I0]
  auto tv3 = add(tv0_cache, tv2); // output
  fusion.addOutput(tv3);

  // I get the same error regardless of this condition:
  // Merging IterDomains requires that their iteration types match. Outer: iS9{i0}, Inner: rS10{i3}
  if (true) {
    tv1_cache->merge(0, 1)->split(0, 4);
    // tv1_cache: [I0*I1/4, 4]

    scheduler_utils::transformPropagateToAllFrom(tv1_cache, -1);

    tv1_cache->axis(-1)->parallelize(ParallelType::TIDx);
    tv1_cache->axis(-2)->parallelize(ParallelType::BIDx);

    scheduler_utils::parallelizeAllLike(tv1_cache);

    fusion.printTransforms();
  } else {
    IdModel id_model(&fusion);
    ValGraph& graph = id_model.idGraph(IdMappingMode::PERMISSIVE);
    AbstractTensor abten;

    abten.domain.reserve(2);
    const auto addAbstractAxis = [&abten, &graph](IterDomain* id) {
      ValGroup g = graph.toGroup(id);
      abten.domain.emplace_back(ValGroupAndItsGraph{g, &graph});
    };
    addAbstractAxis(tv1_cache->axis(0));
    addAbstractAxis(tv1_cache->axis(1));

    abten.merge(0, 1);
    abten.split(0, 4);

    for (TensorView* tv : {tv0_cache, tv1_cache, tv2, tv3}) {
      AbstractTensor local_abten = forwardAroundMissingAxes(abten, tv);
      applyAbstractTransforms(local_abten, tv, &graph);

      // Check that every tensor is scheduled with two loop dimensions and has
      // inner dimension 4 resulting from a Split
      EXPECT_EQ(tv->nDims(), 2);
      EXPECT_TRUE(tv->axis(-1)->extent()->isConst());
      EXPECT_EQ(tv->axis(-1)->extent()->value().as<int64_t>(), 4);
      EXPECT_TRUE(tv->axis(-1)->definition()->isA<Split>());

      tv->axis(-1)->parallelize(ParallelType::TIDx);
      tv->axis(-2)->parallelize(ParallelType::BIDx);
    }
  }

  inlineMost();

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  at::Tensor t0 = at::randn({24}, options);
  at::Tensor t1 = at::randn({24, 1024}, options);
  std::vector<c10::IValue> inputs{t0, t1};

  FusionExecutor fe;
  fe.compileFusion(&fusion, inputs);
  auto cg_outputs = fe.runFusion(inputs);

  testValidate(&fusion, cg_outputs, inputs, __LINE__, __FILE__);
}
```

I agree with an incremental approach and I'm fine with taking out `forwardAroundMissingAxes` and looking at it in a separate PR. I do think we'll want to have it eventually though since for example in our existing test we would need to manually schedule three separate `AbstractTensor`s in order to do the scheduling, at which point we might as well just write a function that schedules a concrete `TensorView` and apply that directly instead of using an `AbstractTensor`.